bee-network/bee-autopeering/src/task.rs

```diff
@@ -0,0 +1,150 @@
+    time::SECOND,
+};
+
+use priority_queue::PriorityQueue;
+use tokio::{sync::oneshot, task::JoinHandle, time};
+
+use std::{collections::HashMap, future::Future, time::Duration};
+
+pub(crate) const MAX_SHUTDOWN_PRIORITY: u8 = 255;
+const SHUTDOWN_TIMEOUT_SECS: Duration = Duration::from_secs(5 * SECOND);
+
+pub(crate) type ShutdownRx = oneshot::Receiver<()>;
+type ShutdownTx = oneshot::Sender<()>;
+
+pub(crate) type Repeat<T> = Box<dyn for<'a> Fn(&'a T) + Send>;
+
+/// Represents types driving an event loop.
+#[async_trait::async_trait]
+pub(crate) trait Runnable {
+    const NAME: &'static str;
+    const SHUTDOWN_PRIORITY: u8;
+
+    type ShutdownSignal: Future + Send + Unpin + 'static;
+
+    async fn run(self, shutdown_rx: Self::ShutdownSignal);
+}
+
+pub(crate) struct TaskManager<S: PeerStore, const N: usize> {
+    shutdown_handles: HashMap<String, JoinHandle<()>>,
+    shutdown_order: PriorityQueue<String, u8>,
+    shutdown_senders: HashMap<String, ShutdownTx>,
+    peerstore: S,
+    active_peers: ActivePeersList,
+    replacements: ReplacementList,
+}
+
+impl<S: PeerStore, const N: usize> TaskManager<S, N> {
+    pub(crate) fn new(peerstore: S, active_peers: ActivePeersList, replacements: ReplacementList) -> Self {
+        Self {
+            shutdown_order: PriorityQueue::with_capacity(N),
+            shutdown_senders: HashMap::with_capacity(N),
+            shutdown_handles: HashMap::with_capacity(N),
+            peerstore,
+            active_peers,
+            replacements,
+        }
+    }
+
+    /// Runs a `Runnable`, which is a type that features an event loop.
+    pub(crate) fn run<R>(&mut self, runnable: R)
+    where
+        R: Runnable<ShutdownSignal = ShutdownRx> + 'static,
+    {
+        let (shutdown_tx, shutdown_rx) = oneshot::channel::<()>();
+        self.shutdown_senders.insert(R::NAME.into(), shutdown_tx);
+
+        let handle = tokio::spawn(runnable.run(shutdown_rx));
+        log::trace!("`{}` running.", R::NAME);
+
+        assert!(!self.shutdown_handles.contains_key(R::NAME));
+        self.shutdown_handles.insert(R::NAME.into(), handle);
+
+        self.shutdown_order.push(R::NAME.into(), R::SHUTDOWN_PRIORITY);
+    }
+
+    /// Repeats a command in certain intervals provided a context `T`. Will be shut down gracefully with the rest of
+    /// all spawned tasks by specifying a `name` and a `shutdown_priority`.
+    pub(crate) fn repeat<T, D>(&mut self, cmd: Repeat<T>, mut delay: D, ctx: T, name: &str, shutdown_priority: u8)
+    where
+        T: Send + Sync + 'static,
+        D: Iterator<Item = Duration> + Send + 'static,
+    {
+        let (shutdown_tx, mut shutdown_rx) = oneshot::channel::<()>();
+        self.shutdown_senders.insert(name.into(), shutdown_tx);
+
+        let handle = tokio::spawn(async move {
+            for duration in &mut delay {
+                tokio::select! {
+                    _ = &mut shutdown_rx => break,
+                    _ = time::sleep(duration) => cmd(&ctx),
```

Comment:
Should we use `interval` instead ? I'm a bit confused by what they mean in the documentation with `No work is performed while awaiting on the sleep future to complete.`
https://docs.rs/tokio/1.14.0/tokio/time/fn.sleep.html

Reply:
`interval` unfortunately only works with fixed intervals.

I am confused by the comment as well. It probably means, that `sleep` doesn't do anything behind the scenes (doesn't drain cpu cycles)

Reply:
Closing this because `interval` requires the duration/delay to be known upfront.